dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+
+    public async Task<TQueryResult> QueryAsync<TQueryResult>(
+      Func<TQueryResult> query, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<T, TQueryResult>(
+      Func<T, TQueryResult> query, T arg, RPCOptions? options = null) { /*...*/ }
+
+    public async Task CancelAsync(RPCOptions? options = null) { /*...*/ }
+
+    public async Task TerminateAsync(
+        string reason? = null, object[]? args = null, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<WorkflowExecutionDescription> DescribeAsync(RPCOptions? options = null) { /*...*/ }
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    public async Task<WorkflowHistory> FetchHistoryAsync(
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
+        bool skipArchival = false,
+        RPCOptions? options = null) { /*...*/ }
+
+    public IAsyncEnumerator<HistoryEvent> FetchHistoryEvents(
+        bool waitNewEvent = false,
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
+        bool skipArchival = false,
+        RPCOptions? options = null) { /*...*/ }
+
+#endif
+}
+```
+
+Notes:
+
+* Notice you can provide a return type explicitly on the `GetResultAsync` call if the generic one on the handle isn't
+  what you want
+* Like start workflow, signal/query string-based calls accept an array of objects instead of trying to use varargs or
+  separating single arg vs multi. This makes the overload count better and will give users compile time errors if it's
+  not literally an object array
+* Notice that `TerminateAsync` does not _require_ `args` like start, signal, and query calls that accept `args` arrays
+  because we don't think it's important for users to explicitly send empty if they don't want them like we think they
+  should for the others
+* Same issue with `FetchHistory*` as `ListWorkflows` on client - no `IAsyncEnumerator` on really old .NET versions, but
+  we're ok with that
+
+## Worker
+
+Usage:
+
+```csharp
+using Temporal.Client;
+using Temporal.Worker;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Build worker options
+var options = new TemporalWorkerOptions("task queue");
+options.AddActivities(new GreetingActivities());
+
+// Run until Ctrl+C
+var worker = new TemporalWorker(client, options);
+var cts = new CancellationTokenSource();
+Console.CancelKeyPress += (sender, eventArgs) => {
+    eventArgs.Cancel = true;
+    cts.Cancel();
+};
+await worker.ExecuteAsync(cts.Token);
+```
+
+Looks like:
+
+```csharp
+namespace Temporal.Worker;
+
+using Temporal.Activity;
+
+public class TemporalWorkerOptions
+{
+    public TemporalWorkerOptions() { /*...*/ }
+    public TemporalWorkerOptions(string taskQueue) { /*...*/ }
+
+    public string? TaskQueue { get; set; }
+
+    public List<ActivityDefinition> ActivityDefinitions { get; set; } = new();
```

Comment:
Can you provide more info on what's an ActivityDefinition and when a user might want to add a definition directly?

Reply:
In the notes I have:

> Activity definitions are the underlying type of the activity collection (they have just a name and method to call)

It's harmless to expose this in the same way as you could call `temporalio.activity.defn(name="foo")(myFunction)` in Python. I can't think of how it might be used, but with these options classes, I usually see in .NET the raw underlying type with helpers to add. You don't usually make things private in options classes.